File: server/python_microservice/session_monitor.py
#!/usr/bin/env python3
"""
Session Surveillance Microservice
Handles real-time session monitoring, anti-cheating detection, and alert generation
"""

import asyncio
import json
import logging
import os
import time
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, asdict
from enum import Enum

import redis
import pymongo
from fastapi import FastAPI, WebSocket, WebSocketDisconnect, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel, Field
import uvicorn
from motor.motor_asyncio import AsyncIOMotorClient
import numpy as np
from sklearn.ensemble import IsolationForest
import joblib

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

# Configuration
REDIS_URL = os.getenv('REDIS_URL', 'redis://localhost:6379')
MONGODB_URL = os.getenv('MONGODB_URL', 'mongodb://localhost:27017')
NODEJS_BACKEND_URL = os.getenv('NODEJS_BACKEND_URL', 'http://localhost:5000')

# Thresholds
TAB_SWITCH_THRESHOLD = 3  # Max tab switches per minute
INACTIVITY_THRESHOLD = 30  # Seconds of inactivity
HEARTBEAT_TIMEOUT = 10  # Seconds without heartbeat
SUSPICIOUS_SCORE_THRESHOLD = 0.7  # ML model threshold

class EventType(Enum):
    TAB_SWITCH = "tab_switch"
    INACTIVITY = "inactivity"
    HEARTBEAT = "heartbeat"
    SESSION_START = "session_start"
    SESSION_END = "session_end"
    SCREEN_LOCK = "screen_lock"
    DEVICE_CHANGE = "device_change"
    COPY_PASTE = "copy_paste"
    RIGHT_CLICK = "right_click"
    KEYBOARD_SHORTCUT = "keyboard_shortcut"

class SecurityLevel(Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

@dataclass
class SessionEvent:
    user_id: str
    session_id: str
    event_type: EventType
    timestamp: datetime
    metadata: Dict[str, Any]
    device_fingerprint: Optional[str] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None

@dataclass
class SecurityAlert:
    alert_id: str
    user_id: str
    session_id: str
    security_level: SecurityLevel
    event_type: EventType
    description: str
    timestamp: datetime
    metadata: Dict[str, Any]
    is_resolved: bool = False

class EventData(BaseModel):
    user_id: str
    session_id: str
    event_type: str
    metadata: Dict[str, Any] = Field(default_factory=dict)
    device_fingerprint: Optional[str] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None

class SessionMonitor:
    def __init__(self):
        self.app = FastAPI(title="Session Surveillance Microservice")
        self.redis_client = redis.Redis.from_url(REDIS_URL, decode_responses=True)
        self.mongo_client = AsyncIOMotorClient(MONGODB_URL)
        self.db = self.mongo_client.session_monitoring
        self.active_sessions: Dict[str, Dict] = {}
        self.ml_model = self._initialize_ml_model()
        
        # Setup CORS
        self.app.add_middleware(
            CORSMiddleware,
            allow_origins=["*"],
            allow_credentials=True,
            allow_methods=["*"],
            allow_headers=["*"],
        )
        
        # Setup routes
        self._setup_routes()
        
    def _initialize_ml_model(self):
        """Initialize ML model for anomaly detection"""
        try:
            # Load pre-trained model if exists
            model_path = "models/anomaly_detector.pkl"
            if os.path.exists(model_path):
                return joblib.load(model_path)
            else:
                # Initialize new model
                model = IsolationForest(contamination=0.1, random_state=42)
                return model
        except Exception as e:
            logger.error(f"Failed to initialize ML model: {e}")
            return None

    def _setup_routes(self):
        """Setup FastAPI routes"""
        
        @self.app.websocket("/ws/session-monitor")
        async def websocket_endpoint(websocket: WebSocket):
            await websocket.accept()
            try:
                while True:
                    data = await websocket.receive_text()
                    event_data = json.loads(data)
                    await self.process_event(event_data)
            except WebSocketDisconnect:
                logger.info("WebSocket disconnected")
            except Exception as e:
                logger.error(f"WebSocket error: {e}")

        @self.app.post("/api/events")
        async def receive_event(event: EventData):
            """Receive events via REST API"""
            try:
                await self.process_event(event.dict())
                return {"status": "success", "message": "Event processed"}
            except Exception as e:
                logger.error(f"Error processing event: {e}")
                raise HTTPException(status_code=500, detail=str(e))

        @self.app.get("/api/sessions/{session_id}/status")
        async def get_session_status(session_id: str):
            """Get session status and alerts"""
            session = self.active_sessions.get(session_id)
            if not session:
                raise HTTPException(status_code=404, detail="Session not found")
            
            alerts = await self.get_session_alerts(session_id)
            return {
                "session_id": session_id,
                "status": session.get("status", "unknown"),
                "last_heartbeat": session.get("last_heartbeat"),
                "alerts": alerts,
                "risk_score": session.get("risk_score", 0.0)
            }

        @self.app.get("/api/admin/sessions")
        async def get_all_sessions():
            """Get all active sessions for admin dashboard"""
            sessions = []
            for session_id, session_data in self.active_sessions.items():
                alerts = await self.get_session_alerts(session_id)
                sessions.append({
                    "session_id": session_id,
                    "user_id": session_data.get("user_id"),
                    "status": session_data.get("status"),
                    "start_time": session_data.get("start_time"),
                    "last_heartbeat": session_data.get("last_heartbeat"),
                    "risk_score": session_data.get("risk_score", 0.0),
                    "alert_count": len(alerts)
                })
            return {"sessions": sessions}

        @self.app.get("/health")
        async def health_check():
            return {"status": "ok"}

    async def process_event(self, event_data: Dict[str, Any]):
        """Process incoming session events"""
        try:
            # Create session event
            event = SessionEvent(
                user_id=event_data["user_id"],
                session_id=event_data["session_id"],
                event_type=EventType(event_data["event_type"]),
                timestamp=datetime.utcnow(),
                metadata=event_data.get("metadata", {}),
                device_fingerprint=event_data.get("device_fingerprint"),
                ip_address=event_data.get("ip_address"),
                user_agent=event_data.get("user_agent")
            )
            
            # Store event in MongoDB
            await self.store_event(event)
            
            # Update active sessions
            await self.update_session_state(event)
            
            # Analyze for suspicious activity
            await self.analyze_security(event)
            
            # Publish to Redis for real-time updates
            await self.publish_event(event)
            
        except Exception as e:
            logger.error(f"Error processing event: {e}")

    async def store_event(self, event: SessionEvent):
        """Store event in MongoDB"""
        try:
            event_doc = {
                "user_id": event.user_id,
                "session_id": event.session_id,
                "event_type": event.event_type.value,
                "timestamp": event.timestamp,
                "metadata": event.metadata,
                "device_fingerprint": event.device_fingerprint,
                "ip_address": event.ip_address,
                "user_agent": event.user_agent
            }
            
            await self.db.session_events.insert_one(event_doc)
            logger.info(f"Stored event: {event.event_type.value} for session {event.session_id}")
            
        except Exception as e:
            logger.error(f"Error storing event: {e}")

    async def update_session_state(self, event: SessionEvent):
        """Update active session state"""
        session_id = event.session_id
        
        if event.event_type == EventType.SESSION_START:
            self.active_sessions[session_id] = {
                "user_id": event.user_id,
                "start_time": event.timestamp,
                "last_heartbeat": event.timestamp,
                "status": "active",
                "risk_score": 0.0,
                "event_count": 0,
                "tab_switches": 0,
                "last_tab_switch": None
            }
            
        elif event.event_type == EventType.HEARTBEAT:
            if session_id in self.active_sessions:
                self.active_sessions[session_id]["last_heartbeat"] = event.timestamp
                
        elif event.event_type == EventType.SESSION_END:
            if session_id in self.active_sessions:
                self.active_sessions[session_id]["status"] = "ended"
                
        elif event.event_type == EventType.TAB_SWITCH:
            if session_id in self.active_sessions:
                session = self.active_sessions[session_id]
                session["tab_switches"] += 1
                session["last_tab_switch"] = event.timestamp
                
        # Update event count
        if session_id in self.active_sessions:
            self.active_sessions[session_id]["event_count"] += 1

    async def analyze_security(self, event: SessionEvent):
        """Analyze event for security threats"""
        session_id = event.session_id
        session = self.active_sessions.get(session_id)
        
        if not session:
            return
            
        alerts = []
        
        # Check for excessive tab switches
        if event.event_type == EventType.TAB_SWITCH:
            time_window = timedelta(minutes=1)
            if session["last_tab_switch"]:
                time_diff = event.timestamp - session["last_tab_switch"]
                if time_diff < time_window and session["tab_switches"] > TAB_SWITCH_THRESHOLD:
                    alerts.append(self._create_alert(
                        event, SecurityLevel.HIGH,
                        f"Excessive tab switching detected: {session['tab_switches']} switches in 1 minute"
                    ))
        
        # Check for inactivity
        if event.event_type == EventType.INACTIVITY:
            if event.metadata.get("duration", 0) > INACTIVITY_THRESHOLD:
                alerts.append(self._create_alert(
                    event, SecurityLevel.MEDIUM,
                    f"User inactive for {event.metadata.get('duration')} seconds"
                ))
        
        # Check for heartbeat timeout
        if session["last_heartbeat"]:
            time_since_heartbeat = datetime.utcnow() - session["last_heartbeat"]
            if time_since_heartbeat.total_seconds() > HEARTBEAT_TIMEOUT:
                alerts.append(self._create_alert(
                    event, SecurityLevel.CRITICAL,
                    f"Heartbeat timeout: {time_since_heartbeat.total_seconds()} seconds"
                ))
        
        # ML-based anomaly detection
        if self.ml_model:
            risk_score = await self._calculate_ml_risk_score(event, session)
            session["risk_score"] = risk_score
            
            if risk_score > SUSPICIOUS_SCORE_THRESHOLD:
                alerts.append(self._create_alert(
                    event, SecurityLevel.HIGH,
                    f"ML model flagged suspicious activity (score: {risk_score:.2f})"
                ))
        
        # Store alerts
        for alert in alerts:
            await self.store_alert(alert)
            
        # Update session risk score
        if alerts:
            session["risk_score"] = max(session["risk_score"], 0.8)

    def _create_alert(self, event: SessionEvent, level: SecurityLevel, description: str) -> SecurityAlert:
        """Create security alert"""
        return SecurityAlert(
            alert_id=f"alert_{int(time.time())}_{event.session_id}",
            user_id=event.user_id,
            session_id=event.session_id,
            security_level=level,
            event_type=event.event_type,
            description=description,
            timestamp=event.timestamp,
            metadata=event.metadata
        )

    async def _calculate_ml_risk_score(self, event: SessionEvent, session: Dict) -> float:
        """Calculate ML-based risk score"""
        try:
            # Extract features for ML model
            features = [
                session.get("event_count", 0),
                session.get("tab_switches", 0),
                session.get("risk_score", 0.0),
                event.metadata.get("duration", 0),
                event.metadata.get("click_count", 0),
                event.metadata.get("keypress_count", 0)
            ]
            
            # Normalize features
            features = np.array(features).reshape(1, -1)
            
            # Get anomaly score (lower = more anomalous)
            if self.ml_model:
                score = self.ml_model.decision_function(features)[0]
                # Convert to risk score (0-1, higher = more risky)
                risk_score = 1 - (score + 0.5)  # Normalize to 0-1
                return max(0.0, min(1.0, risk_score))
            
        except Exception as e:
            logger.error(f"Error calculating ML risk score: {e}")
            
        return 0.0

    async def store_alert(self, alert: SecurityAlert):
        """Store security alert in MongoDB"""
        try:
            alert_doc = asdict(alert)
            alert_doc["security_level"] = alert.security_level.value
            alert_doc["event_type"] = alert.event_type.value
            
            await self.db.security_alerts.insert_one(alert_doc)
            logger.warning(f"Security alert stored: {alert.description}")
            
        except Exception as e:
            logger.error(f"Error storing alert: {e}")

    async def get_session_alerts(self, session_id: str) -> List[Dict]:
        """Get alerts for a specific session"""
        try:
            cursor = self.db.security_alerts.find(
                {"session_id": session_id},
                {"_id": 0}
            ).sort("timestamp", -1).limit(50)
            
            alerts = await cursor.to_list(length=50)
            return alerts
            
        except Exception as e:
            logger.error(f"Error getting session alerts: {e}")
            return []

    async def publish_event(self, event: SessionEvent):
        """Publish event to Redis for real-time updates"""
        try:
            event_data = {
                "user_id": event.user_id,
                "session_id": event.session_id,
                "event_type": event.event_type.value,
                "timestamp": event.timestamp.isoformat(),
                "metadata": event.metadata
            }
            
            # Publish to Redis channel
            self.redis_client.publish("session_events", json.dumps(event_data))
            
        except Exception as e:
            logger.error(f"Error publishing event: {e}")

    async def cleanup_expired_sessions(self):
        """Clean up expired sessions"""
        current_time = datetime.utcnow()
        expired_sessions = []
        
        for session_id, session_data in self.active_sessions.items():
            if session_data.get("status") == "ended":
                expired_sessions.append(session_id)
            elif session_data.get("last_heartbeat"):
                time_since_heartbeat = current_time - session_data["last_heartbeat"]
                if time_since_heartbeat.total_seconds() > 300:  # 5 minutes
                    expired_sessions.append(session_id)
        
        for session_id in expired_sessions:
            del self.active_sessions[session_id]
            logger.info(f"Cleaned up expired session: {session_id}")

    async def start_cleanup_task(self):
        """Start periodic cleanup task"""
        while True:
            await asyncio.sleep(60)  # Run every minute
            await self.cleanup_expired_sessions()

monitor = SessionMonitor()
app = monitor.app

@app.on_event("startup")
async def startup_event():
    asyncio.create_task(monitor.start_cleanup_task()) 

File: server/utils/sessionMonitor.js
import { WebSocketServer } from 'ws';
import Redis from 'ioredis';
import axios from 'axios';
import { logger } from './logger.js';

class SessionMonitor {
  constructor() {
    this.wss = null;
    this.redis = new Redis(process.env.REDIS_URL || 'redis://localhost:6379');
    this.pythonServiceUrl = process.env.PYTHON_SERVICE_URL || 'http://localhost:8001';
    this.activeConnections = new Map(); // sessionId -> WebSocket
    this.sessionData = new Map(); // sessionId -> session info
    this.heartbeatIntervals = new Map(); // sessionId -> interval
    
    // Configuration
    this.HEARTBEAT_INTERVAL = 5000; // 5 seconds
    this.INACTIVITY_TIMEOUT = 30000; // 30 seconds
    this.MAX_TAB_SWITCHES = 3; // per minute
  }

  /**
   * Initialize WebSocket server
   */
  initialize(server) {
    this.wss = new WebSocketServer({ server });
    
    this.wss.on('connection', (ws, req) => {
      this.handleConnection(ws, req);
    });

    // Subscribe to Redis events from Python service
    this.subscribeToRedisEvents();
    
    logger.info('Session Monitor WebSocket server initialized');
  }

  /**
   * Handle new WebSocket connection
   */
  handleConnection(ws, req) {
    const sessionId = this.extractSessionId(req);
    const userId = this.extractUserId(req);
    
    if (!sessionId || !userId) {
      ws.close(1008, 'Missing session or user information');
      return;
    }

    // Store connection
    this.activeConnections.set(sessionId, ws);
    this.sessionData.set(sessionId, {
      userId,
      sessionId,
      connectedAt: new Date(),
      lastHeartbeat: new Date(),
      tabSwitches: 0,
      lastTabSwitch: null,
      status: 'active'
    });

    // Start heartbeat monitoring
    this.startHeartbeatMonitoring(sessionId);

    // Send session start event to Python service
    this.sendEventToPython({
      user_id: userId,
      session_id: sessionId,
      event_type: 'session_start',
      metadata: {
        ip_address: req.socket.remoteAddress,
        user_agent: req.headers['user-agent'],
        connected_at: new Date().toISOString()
      }
    });

    // Handle incoming messages
    ws.on('message', (data) => {
      try {
        const message = JSON.parse(data);
        this.handleMessage(sessionId, message);
      } catch (error) {
        logger.error(`Error parsing message from session ${sessionId}:`, error);
      }
    });

    // Handle connection close
    ws.on('close', () => {
      this.handleDisconnection(sessionId);
    });

    // Handle errors
    ws.on('error', (error) => {
      logger.error(`WebSocket error for session ${sessionId}:`, error);
      this.handleDisconnection(sessionId);
    });

    logger.info(`Session ${sessionId} connected for user ${userId}`);
  }

  /**
   * Handle incoming WebSocket messages
   */
  handleMessage(sessionId, message) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    const { type, data } = message;

    switch (type) {
      case 'heartbeat':
        this.handleHeartbeat(sessionId, data);
        break;
        
      case 'tab_switch':
        this.handleTabSwitch(sessionId, data);
        break;
        
      case 'inactivity':
        this.handleInactivity(sessionId, data);
        break;
        
      case 'screen_lock':
        this.handleScreenLock(sessionId, data);
        break;
        
      case 'copy_paste':
        this.handleCopyPaste(sessionId, data);
        break;
        
      case 'right_click':
        this.handleRightClick(sessionId, data);
        break;
        
      case 'keyboard_shortcut':
        this.handleKeyboardShortcut(sessionId, data);
        break;
        
      case 'device_change':
        this.handleDeviceChange(sessionId, data);
        break;
        
      default:
        logger.warn(`Unknown message type: ${type} from session ${sessionId}`);
    }
  }

  /**
   * Handle heartbeat from frontend
   */
  handleHeartbeat(sessionId, data) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    session.lastHeartbeat = new Date();
    
    // Send heartbeat event to Python service
    this.sendEventToPython({
      user_id: session.userId,
      session_id: sessionId,
      event_type: 'heartbeat',
      metadata: {
        timestamp: new Date().toISOString(),
        ...data
      }
    });
  }

  /**
   * Handle tab switch detection
   */
  handleTabSwitch(sessionId, data) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    const now = new Date();
    session.tabSwitches++;
    
    // Reset counter if more than 1 minute has passed
    if (session.lastTabSwitch) {
      const timeDiff = now - session.lastTabSwitch;
      if (timeDiff > 60000) { // 1 minute
        session.tabSwitches = 1;
      }
    }
    
    session.lastTabSwitch = now;

    // Check for excessive tab switching
    if (session.tabSwitches > this.MAX_TAB_SWITCHES) {
      this.flagSuspiciousActivity(sessionId, 'excessive_tab_switching', {
        tab_switches: session.tabSwitches,
        time_window: '1 minute'
      });
    }

    // Send event to Python service
    this.sendEventToPython({
      user_id: session.userId,
      session_id: sessionId,
      event_type: 'tab_switch',
      metadata: {
        tab_switches: session.tabSwitches,
        timestamp: now.toISOString(),
        ...data
      }
    });
  }

  /**
   * Handle inactivity detection
   */
  handleInactivity(sessionId, data) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    const duration = data.duration || 0;
    
    if (duration > this.INACTIVITY_TIMEOUT) {
      this.flagSuspiciousActivity(sessionId, 'extended_inactivity', {
        duration_seconds: duration
      });
    }

    this.sendEventToPython({
      user_id: session.userId,
      session_id: sessionId,
      event_type: 'inactivity',
      metadata: {
        duration: duration,
        timestamp: new Date().toISOString(),
        ...data
      }
    });
  }

  /**
   * Handle screen lock detection
   */
  handleScreenLock(sessionId, data) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    this.sendEventToPython({
      user_id: session.userId,
      session_id: sessionId,
      event_type: 'screen_lock',
      metadata: {
        timestamp: new Date().toISOString(),
        ...data
      }
    });
  }

  /**
   * Handle copy-paste detection
   */
  handleCopyPaste(sessionId, data) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    this.flagSuspiciousActivity(sessionId, 'copy_paste_detected', data);

    this.sendEventToPython({
      user_id: session.userId,
      session_id: sessionId,
      event_type: 'copy_paste',
      metadata: {
        timestamp: new Date().toISOString(),
        ...data
      }
    });
  }

  /**
   * Handle right-click detection
   */
  handleRightClick(sessionId, data) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    this.sendEventToPython({
      user_id: session.userId,
      session_id: sessionId,
      event_type: 'right_click',
      metadata: {
        timestamp: new Date().toISOString(),
        ...data
      }
    });
  }

  /**
   * Handle keyboard shortcut detection
   */
  handleKeyboardShortcut(sessionId, data) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    this.flagSuspiciousActivity(sessionId, 'keyboard_shortcut_detected', data);

    this.sendEventToPython({
      user_id: session.userId,
      session_id: sessionId,
      event_type: 'keyboard_shortcut',
      metadata: {
        timestamp: new Date().toISOString(),
        ...data
      }
    });
  }

  /**
   * Handle device change detection
   */
  handleDeviceChange(sessionId, data) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    this.flagSuspiciousActivity(sessionId, 'device_change_detected', data);

    this.sendEventToPython({
      user_id: session.userId,
      session_id: sessionId,
      event_type: 'device_change',
      metadata: {
        timestamp: new Date().toISOString(),
        ...data
      }
    });
  }

  /**
   * Flag suspicious activity
   */
  flagSuspiciousActivity(sessionId, activityType, metadata) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    logger.warn(`Suspicious activity detected in session ${sessionId}: ${activityType}`, metadata);

    // Send alert to admin dashboard
    this.sendAlertToAdmin({
      sessionId,
      userId: session.userId,
      activityType,
      metadata,
      timestamp: new Date().toISOString()
    });

    // Optionally lock session
    if (this.shouldLockSession(activityType)) {
      this.lockSession(sessionId, activityType);
    }
  }

  /**
   * Determine if session should be locked
   */
  shouldLockSession(activityType) {
    const criticalActivities = [
      'excessive_tab_switching',
      'copy_paste_detected',
      'keyboard_shortcut_detected',
      'device_change_detected'
    ];
    
    return criticalActivities.includes(activityType);
  }

  /**
   * Lock session
   */
  lockSession(sessionId, reason) {
    const ws = this.activeConnections.get(sessionId);
    if (!ws) return;

    const session = this.sessionData.get(sessionId);
    if (!session) return;

    session.status = 'locked';
    
    // Send lock notification to frontend
    ws.send(JSON.stringify({
      type: 'session_locked',
      data: {
        reason,
        timestamp: new Date().toISOString()
      }
    }));

    logger.warn(`Session ${sessionId} locked due to: ${reason}`);
  }

  /**
   * Start heartbeat monitoring for a session
   */
  startHeartbeatMonitoring(sessionId) {
    const interval = setInterval(() => {
      const session = this.sessionData.get(sessionId);
      if (!session) {
        clearInterval(interval);
        return;
      }

      const now = new Date();
      const timeSinceHeartbeat = now - session.lastHeartbeat;

      if (timeSinceHeartbeat > this.INACTIVITY_TIMEOUT) {
        logger.warn(`Session ${sessionId} heartbeat timeout`);
        this.handleDisconnection(sessionId);
      }
    }, this.HEARTBEAT_INTERVAL);

    this.heartbeatIntervals.set(sessionId, interval);
  }

  /**
   * Handle session disconnection
   */
  handleDisconnection(sessionId) {
    const session = this.sessionData.get(sessionId);
    if (!session) return;

    // Clear heartbeat interval
    const interval = this.heartbeatIntervals.get(sessionId);
    if (interval) {
      clearInterval(interval);
      this.heartbeatIntervals.delete(sessionId);
    }

    // Send session end event to Python service
    this.sendEventToPython({
      user_id: session.userId,
      session_id: sessionId,
      event_type: 'session_end',
      metadata: {
        duration: new Date() - session.connectedAt,
        reason: 'disconnection',
        timestamp: new Date().toISOString()
      }
    });

    // Clean up
    this.activeConnections.delete(sessionId);
    this.sessionData.delete(sessionId);

    logger.info(`Session ${sessionId} disconnected`);
  }

  /**
   * Send event to Python microservice
   */
  async sendEventToPython(eventData) {
    try {
      await axios.post(`${this.pythonServiceUrl}/api/events`, eventData, {
        timeout: 5000,
        headers: {
          'Content-Type': 'application/json'
        }
      });
    } catch (error) {
      logger.error('Failed to send event to Python service:', error.message);
    }
  }

  /**
   * Subscribe to Redis events from Python service
   */
  subscribeToRedisEvents() {
    const subscriber = new Redis(process.env.REDIS_URL || 'redis://localhost:6379');
    
    subscriber.subscribe('session_events', (err) => {
      if (err) {
        logger.error('Failed to subscribe to Redis events:', err);
        return;
      }
      logger.info('Subscribed to session events from Python service');
    });

    subscriber.on('message', (channel, message) => {
      try {
        const event = JSON.parse(message);
        this.handlePythonEvent(event);
      } catch (error) {
        logger.error('Error parsing Redis event:', error);
      }
    });
  }

  /**
   * Handle events from Python service
   */
  handlePythonEvent(event) {
    const { session_id, event_type, metadata } = event;
    
    // Handle different event types
    switch (event_type) {
      case 'security_alert':
        this.handleSecurityAlert(session_id, metadata);
        break;
        
      case 'session_terminated':
        this.terminateSession(session_id, metadata);
        break;
        
      default:
        logger.info(`Received event from Python service: ${event_type}`);
    }
  }

  /**
   * Handle security alerts from Python service
   */
  handleSecurityAlert(sessionId, metadata) {
    const ws = this.activeConnections.get(sessionId);
    if (!ws) return;

    // Send alert to frontend
    ws.send(JSON.stringify({
      type: 'security_alert',
      data: metadata
    }));

    logger.warn(`Security alert for session ${sessionId}:`, metadata);
  }

  /**
   * Terminate session
   */
  terminateSession(sessionId, metadata) {
    const ws = this.activeConnections.get(sessionId);
    if (!ws) return;

    // Send termination notice to frontend
    ws.send(JSON.stringify({
      type: 'session_terminated',
      data: metadata
    }));

    // Close connection
    ws.close(1000, 'Session terminated by admin');

    logger.warn(`Session ${sessionId} terminated:`, metadata);
  }

  /**
   * Send alert to admin dashboard
   */
  sendAlertToAdmin(alertData) {
    // Publish to Redis for admin dashboard
    this.redis.publish('admin_alerts', JSON.stringify(alertData));
  }

  /**
   * Get session status
   */
  getSessionStatus(sessionId) {
    return this.sessionData.get(sessionId);
  }

  /**
   * Get all active sessions
   */
  getAllSessions() {
    const sessions = [];
    for (const [sessionId, session] of this.sessionData) {
      sessions.push({
        sessionId,
        userId: session.userId,
        status: session.status,
        connectedAt: session.connectedAt,
        lastHeartbeat: session.lastHeartbeat,
        tabSwitches: session.tabSwitches
      });
    }
    return sessions;
  }

  /**
   * Extract session ID from request
   */
  extractSessionId(req) {
    const url = new URL(req.url, 'http://localhost');
    return url.searchParams.get('sessionId');
  }

  /**
   * Extract user ID from request
   */
  extractUserId(req) {
    const url = new URL(req.url, 'http://localhost');
    return url.searchParams.get('userId');
  }

  /**
   * Broadcast message to all connected sessions
   */
  broadcast(message) {
    this.wss.clients.forEach((client) => {
      if (client.readyState === WebSocketServer.OPEN) {
        client.send(JSON.stringify(message));
      }
    });
  }

  /**
   * Send message to specific session
   */
  sendToSession(sessionId, message) {
    const ws = this.activeConnections.get(sessionId);
    if (ws && ws.readyState === WebSocketServer.OPEN) {
      ws.send(JSON.stringify(message));
    }
  }
}

// Create singleton instance
const sessionMonitor = new SessionMonitor();

export default sessionMonitor; 